### ISSUE TITLE
[@typescript-eslint/no-shadow] Does not report when var is after function

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3605]
**Repro**

```TS
function f(x: unknown): void {
  console.log(x);
}

const x = 1;
```

**Expected Result**

Should report issue:
'x' is already declared in the upper scope

**Actual Result**

Does not catch issue.

**Additional Info**

The rule works if you move constant before function.
```TS
const x = 1;

function f(x: unknown): void {
  console.log(x);
}
```

However, typescript allows any relative position. So, the rule should not rely on variable being defined before the function.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.1` |
| `@typescript-eslint/parser`        | `4.28.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.30.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
this is working as expected. See the base rule:
https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogZXJyb3IgKi9cblxuZnVuY3Rpb24gZih4KSB7XG4gIGNvbnNvbGUubG9nKHgpO1xufVxuXG5jb25zdCB4ID0gMTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19
